src/compiler/factory/emitHelpers.ts

```diff
@@ -1422,6 +1438,21 @@ const disposeResourcesHelper: UnscopedEmitHelper = {
         });`,
 };
 
+const rewriteRelativeImportExtensionsHelper: UnscopedEmitHelper = {
+    name: "typescript:rewriteRelativeImportExtensions",
+    importName: "__rewriteRelativeImportExtension",
+    scoped: false,
+    text: `
+        var __rewriteRelativeImportExtension = (this && this.__rewriteRelativeImportExtension) || function (path, preserveJsx) {
+            if (typeof path === "string" && /^\\.\\.?\\//.test(path)) {
+                return path.replace(/\\.(tsx)$|((?:\\.d)?)((?:\\.[^./]+?)?)\\.([cm]?)ts$/i, function (m, tsx, d, ext, cm) {

```

Comment:
If a path contains `.jsx` it is preserved even with the `preserveJsx` argument. 

Should the regex be `/\.([jt]sx)$|((?:\.d)?)((?:\.[^./]+?)?)\.([cm]?)ts$/i`?

This now works:

```ts
console.log(rewriteRelativeImportExtension("./main.tsx"));
console.log(rewriteRelativeImportExtension("./main.ts"));
console.log(rewriteRelativeImportExtension("./main.jsx"));
console.log(rewriteRelativeImportExtension("./main.jsx", true));
console.log(rewriteRelativeImportExtension("./main.js"));
```

```
./main.js
./main.js
./main.js
./main.jsx
./main.js
```

Reply:
Can you clarify your actual behavior and expected behavior?

Reply:
Nevermind, I'm writing a plugin that is based on `__rewriteRelativeImportExtension`. For Typescript is fine, however for esbuild the `.jsx` can the transformed to `.js`.